From 1dc672fe1c4786d0ac8c32bacf8bbef78b2660c3 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Tue, 10 Jan 2017 12:49:04 +0100
Subject: [PATCH 0748/1345] ARM: mvebu: Enable WA for setting frequency ID on
 AXP during CPU_PM_EXIT

commit  b4d58ace80029b86a728f9bddb0eef0ea0d41303 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Commit 8d9157000305 ("ARM: mvebu: Armada 38x: Add dynamic frequency
scaling support in pmsu") extended existing cpufreq support for Armada XP.
However it appeared that modifying Frequency ID for this SoC during
CPU_PM_EXIT, which was introduced by the commit, results in hang during
boot.

This patch enables WA for above situation on Armada XP by introducing
a dedicated flag, which allows for conditional modification of Frequency
ID field in mvebu_v7_pmsu_idle_exit().

Partial fix 2/2 (more a WA) for JIRA SYSTEMSW-3238
("Armada XP fails to boot in the LSP v4.4.8")

Change-Id: I08f9cf0b4e203dedf30be7ffc5fa5d6650e527f5
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Signed-off-by: Igal Liberman <igall@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35869
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-mvebu/pmsu.c |   27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-mvebu/pmsu.c b/arch/arm/mach-mvebu/pmsu.c
index a8e9218..2b77422 100644
--- a/arch/arm/mach-mvebu/pmsu.c
+++ b/arch/arm/mach-mvebu/pmsu.c
@@ -114,6 +114,8 @@
 static void __iomem *cib_control;
 static void __iomem *pmsu_msys_ba_redirect_reg;
 
+static bool freq_id_wa;
+
 static void *mvebu_cpu_resume;
 int (*mvebu_pmsu_dfs_request_ptr)(int cpu);
 
@@ -262,6 +264,16 @@ static int __init mvebu_v7_pmsu_init(void)
 			iounmap(pmsu_mp_base);
 	}
 
+	if (of_machine_is_compatible("marvell,armadaxp")) {
+		/*
+		 * Enable work-around for modifying Frequency ID during
+		 * CPU_PM_EXIT. Change added originally for Armada 38x
+		 * SoC does not work on Armada XP and causes very early
+		 * fail during boot.
+		 */
+		freq_id_wa = true;
+	}
+
  out:
 	of_node_put(np);
 	return ret;
@@ -413,11 +425,18 @@ void mvebu_v7_pmsu_idle_exit(void)
 	reg &= ~PMSU_CTL_CFG_L2_PWDDN;
 
 	/*
-	 * When exiting from idle state such as cpuidle or hotplug,
-	 * Enable PMU wait for the CPU to enter WFI when doing DFS
-	 * by setting CPUx Frequency ID to 1
+	 * Below modification of Frequency ID is valid only on Armada 38x
+	 * SoC. Due to hardware issue it causes fail on Armada XP, so
+	 * avoid this section, depending on freq_id_wa flag value.
 	 */
-	reg |= 1 << PMSU_CTL_CFG_CPU0_FRQ_ID_SFT;
+	if (!freq_id_wa) {
+		/*
+		 * When exiting from idle state such as cpuidle or hotplug,
+		 * Enable PMU wait for the CPU to enter WFI when doing DFS
+		 * by setting CPUx Frequency ID to 1
+		 */
+		reg |= 1 << PMSU_CTL_CFG_CPU0_FRQ_ID_SFT;
+	}
 	writel(reg, pmsu_mp_base + PMSU_CTL_CFG(hw_cpu));
 
 	/* cancel Enable wakeup events and mask interrupts */
-- 
1.7.9.5

